packages/assets-controllers/src/TokensController.ts

```diff
@@ -730,28 +730,110 @@ export class TokensController extends BaseControllerV1<
       throw new Error(`Asset of type ${type} not supported`);
     }
 
-    const { selectedAddress } = this.config;
+    if (!asset.address) {
+      throw rpcErrors.invalidParams('Address must be specified');
+    }
+
+    if (!isValidHexAddress(asset.address)) {
+      throw rpcErrors.invalidParams(`Invalid address "${asset.address}"`);
+    }
+
+    // Validate contract
+
+    if (await this._detectIsERC721(asset.address, networkClientId)) {
+      throw rpcErrors.invalidParams(
+        `Contract ${asset.address} must match type ${type}, but was detected as ${ERC721}`,
+      );
+    }
+
+    const provider = this._getProvider(networkClientId);
+    const isErc1155 = await safelyExecute(() =>
+      new ERC1155Standard(provider).contractSupportsBase1155Interface(
+        asset.address,
+      ),
+    );
+    if (isErc1155) {
+      throw rpcErrors.invalidParams(
+        `Contract ${asset.address} must match type ${type}, but was detected as ${ERC1155}`,
+      );
+    }
+
+    const erc20 = new ERC20Standard(provider);
+    const [contractName, contractSymbol, contractDecimals] = await Promise.all([
+      safelyExecute(() => erc20.getTokenName(asset.address)),
+      safelyExecute(() => erc20.getTokenSymbol(asset.address)),
+      safelyExecute(async () => erc20.getTokenDecimals(asset.address)),
+    ]);
+
+    asset.name = contractName;
+
+    // Validate symbol
+
+    if (!asset.symbol && !contractSymbol) {
+      throw rpcErrors.invalidParams(
+        'A symbol is required, but was not found in the request or contract',
+      );
+    }
+
+    if (
+      contractSymbol !== undefined &&
+      asset.symbol !== undefined &&
+      asset.symbol !== contractSymbol
+    ) {
+      throw rpcErrors.invalidParams(
+        `The symbol in the request (${asset.symbol}) does not match the symbol in the contract (${contractSymbol})`,
```

Comment:
IMO if it's in both the request and contract, it's nicer to validate they match vs overwrite with what's on-chain. Would be confusing if a dapp requests something, and we watch something different. Nice to have user, dapp, and chain in agreement.

Only concern is breaking existing code with the stronger validation. e.g. symbol is "ABC" on-chain, but they watch "abc" or "$ABC". If I find evidence of this, the comparison can be adjusted.

Reply:
could do a case insensitive check just to be safe

Reply:
Made symbol comparison case insensitive with test here: https://github.com/MetaMask/core/pull/1745/commits/c5b3d34d132da7ce48bd51f1fa81a2bf2a5c753e